src/FrameSemantics.cc

```diff
@@ -282,289 +282,511 @@ static Errors resolveModelPoseWithPlacementFrame(
-  edge.SetWeight(0);
+  /// \brief Raw pose of the entity.
+  const ignition::math::Pose3d rawPose;
+  /// \brief The //pose/@relative_to attribute.
+  const std::string relativeTo;
+  /// \brief The //frame/@attached_to attribute.
+  const std::string attachedTo;
+  /// \brief The final @relative_to attribute. For sdf::Frame, this is set to
+  /// the attachedTo if the relativeTo is empty. For sdf::InterfaceFrame, it's
+  /// always set to attachedTo.
+  const std::string finalRelativeTo;
+};
 
-  // add link vertices
-  for (uint64_t l = 0; l < _model->LinkCount(); ++l)
+/// \brief Wrapper for sdf::Joint and sdf::InterfaceJoint
+struct JointWrapper: public WrapperBase
+{
+  /// \brief Constructor that takes an sdf::Joint
+  explicit JointWrapper(const sdf::Joint &_joint)
+      : WrapperBase{_joint.Name(), "Joint", FrameType::JOINT},
+        rawPose(_joint.RawPose()),
+        relativeTo(_joint.PoseRelativeTo()),
+        childName(_joint.ChildLinkName()),
+        finalRelativeTo(relativeTo.empty() ? childName : relativeTo)
+  {
+  }
+
+  /// \brief Constructor that takes an sdf::InterfaceJoint
+  explicit JointWrapper(const sdf::InterfaceJoint &_ifaceJoint)
+      : WrapperBase{_ifaceJoint.Name(), "Interface Joint", FrameType::JOINT},
+        rawPose(_ifaceJoint.PoseInChildFrame()),
+        relativeTo(_ifaceJoint.ChildName()),
+        childName(_ifaceJoint.ChildName()),
+        finalRelativeTo(childName)
+  {
+  }
+
+  /// \brief Raw pose of the entity.
+  const ignition::math::Pose3d rawPose;
+  /// \brief The //pose/@relative_to attribute.
+  const std::string relativeTo;
+  /// \brief The name of the child frame.
+  const std::string childName;
+  /// \brief The final @relative_to attribute. For sdf::Joint, this is set to
+  /// the childName if the relativeTo. For sdf::InterfaceJoint, it's always set
+  /// to childName.
+  const std::string finalRelativeTo;
+};
+
+/// \brief Wrapper for sdf::Model and sdf::InterfaceModel
+struct ModelWrapper: public WrapperBase
+{
+  /// \brief Constructor that takes an sdf::Model
+  explicit ModelWrapper(const sdf::Model &_model)
+      : WrapperBase{_model.Name(), "Model",
+                    _model.Static() ? FrameType::STATIC_MODEL
+                                    : FrameType::MODEL},
+        rawPose(_model.RawPose()),
+        relativeTo(_model.PoseRelativeTo()),
+        finalRelativeTo(relativeTo),
+        canonicalLinkName(_model.CanonicalLinkAndRelativeName().second),
+        placementFrameName(_model.PlacementFrameName()),
+        isStatic(_model.Static())
   {
-    auto link = _model->LinkByIndex(l);
-    if (outModel.Count(link->Name()) > 0)
+    for (uint64_t i = 0; i < _model.LinkCount(); ++i)
     {
-      errors.push_back({ErrorCode::DUPLICATE_NAME,
-          "Link with non-unique name [" + link->Name() +
-          "] detected in model with name [" + _model->Name() +
-          "]."});
-      continue;
+      links.emplace_back(*_model.LinkByIndex(i));
```

Comment:
nit (and below too):
```suggestion
      this->links.emplace_back(*_model.LinkByIndex(i));
```

Reply:
Done in https://github.com/ignitionrobotics/sdformat/pull/764/commits/59050f96134d6bf0cea07e35ab7916b1da7b70d9.